<?xml version="1.0"?>
<robot xmlns:sensor="http://playerstage.sourceforge.net/gazebo/xmlschema/#sensor"
       xmlns:controller="http://playerstage.sourceforge.net/gazebo/xmlschema/#controller"
       xmlns:interface="http://playerstage.sourceforge.net/gazebo/xmlschema/#interface">

  <property name="M_PI" value="3.1415926535897931" />
 
  <macro name="m3_r_palm" params="side parent reflect *origin">

  <joint name="palm_joint" type="fixed">
    <origin xyz="0 0 -0.04414" rpy="0 0 0" />
    <parent link="${parent}"/>
    <child link="palm" />
  </joint>

     <link name="palm">
      <inertial>
        <mass value="0.663" />
	<origin xyz="0.001459 0.001619 -0.051312"/>
	<inertia ixx="0.000873"
	iyy="0.000597"
	izz="0.000565"
	ixy="0.000031"
	ixz="-0.000008"
	iyz="0.000016" />
      </inertial>

      <visual>
	      <material name="GreyA" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="palm_geom">
		      <mesh filename="package://${MESH}/H2R3/H2R3_RT8_palm_right.dae" />
	      </geometry>  
      </visual>

      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="palm_collision">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_RT8_palm_right.stl" />
	      </geometry>
      </collision>
    </link>

<!-- Start Thumb -->
    <joint name="right_hand_j0" type="revolute">
      <axis xyz="0 0 -1" />
      <origin xyz="-0.02 0.028 -0.03" rpy="0 0 0" />
      <parent link="palm"/>
      <child link="thumb0"/>
      <limit effort="100" velocity="${20.0*M_PI/180}" lower="${-40*M_PI/180}" upper="${15*M_PI/180}" />
    </joint>

    <link name="thumb0">
      <inertial>
        <mass value="0.0117" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>
      <visual>
	      <material name="GreyA" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="thumb0_geom">
		      <mesh filename="package://${MESH}/H2R3/H2R3_RJ0_thumb_abductlink.dae" />
	      </geometry>  
      </visual>
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="thumb0_collision">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_RJ0_thumb_abductlink.stl" />
	      </geometry>
      </collision>
    </link>
  
    <joint name="right_hand_j1" type="revolute">
      <axis xyz="0 -1 0"/>
     <limit effort="100" velocity="${75*M_PI/180}" lower="${-80.919*M_PI/180}" upper="${80.91*M_PI/180}" />    
      <origin xyz="-0.018 0 0" rpy="0 0 0" />
      <parent link="thumb0"/>
      <child link="thumb1"/>
    </joint>
    <link name="thumb1">
      <inertial>
        <mass value="0.0344" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>
      <visual>
	      <material name="GreyB" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="thumb1_geom">
		      <mesh filename="package://${MESH}/H2R3/H2R3_thumb_knuckle1.dae" />
	      </geometry>  
      </visual>
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="thumb1_collision">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_thumb_knuckle1.stl" />
	      </geometry>
      </collision>
    </link>

    <joint name="right_hand_j2" type="revolute">
      <axis xyz="0 -1 0" />
      <origin xyz="-0.044 0 0" rpy="0 0 0" />
      <parent link="thumb1" />
      <child link="thumb2"/>
      <limit effort="100" velocity="${50*M_PI/180}" lower="${-14.94*M_PI/180}" upper="${14.94*M_PI/180}" />    
    </joint>

    <link name="thumb2">  
      <inertial>
        <mass value="0.01" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>        
      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="thumb_knuckle2_geom">
		      <mesh filename="package://${MESH}/H2R3/H2R3_thumb_knuckle2.dae" />
	      </geometry>  
      </visual>
      
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="thumb_knuckle2_collision">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_thumb_knuckle2.stl" />
	      </geometry>
      </collision>
    </link>

<!-- Start Index Finger -->
    <joint name="right_hand_j3" type="revolute">
      <axis xyz="0 1 0" />
      <origin xyz="0 0.028 -0.101" rpy="0 0 0" />
      <parent link="palm" />
      <child link="index0"/>
	<limit effort="100" velocity="${50*M_PI/180}" lower="${-7.072*M_PI/180}" upper="${37.072*M_PI/180}" />    
    </joint>

    <link name="index0">
      <inertial>
        <mass value="0.01" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>        
      <visual>
	      <material name="GreyA" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="index0_geom">		      
		     <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle1.dae" />
	      </geometry>  
      </visual>      
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="index0__collision">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle1.stl" />
	      </geometry>
      </collision>
    </link>

    <joint name="right_hand_j4" type="revolute">
      <axis xyz="0 1 0" />
      <limit effort="100" velocity="${1600*M_PI/180}" lower="${-34.99*M_PI/180}" upper="${34.99*M_PI/180}" /> 
      <origin xyz="0 0 -0.044" rpy="0 0 0" />
      <parent link="index0"/>
      <child link="index1"/>
    </joint>

    <link name="index1"> 
      <inertial>
        <mass value="0.0429" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>     
      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="index1_geom">		      
		      <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle2.dae" />
	      </geometry>  
      </visual>
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="index1_collision">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle2.stl" />
	      </geometry>
      </collision>
    </link>

    <joint name="right_hand_j5" type="revolute">
      <axis xyz="0 1 0" />
      <limit effort="100" velocity="${1600*M_PI/180}" lower="${-34.21*M_PI/180}" upper="${34.213*M_PI/180}" /> 
      <origin xyz="0 0 -0.032" rpy="0 0 0" />
      <parent link="index1"/>
      <child link="index2"/>
    </joint>

    <link name="index2">
      <inertial>
        <mass value="0.01" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>        
      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="index2_geom">		      
		      <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle3.dae" />
	      </geometry>  
      </visual>
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="index2_collision">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle3.stl" />
	      </geometry>
      </collision>
    </link>

<!-- Start Ring Finger -->
    <joint name="right_hand_j6" type="revolute">
      <axis xyz="0 1 0" />
      <origin xyz="0 0 -0.101" rpy="0 0 0" />
      <parent link="palm" />
      <child link="ring0"/>
	<limit effort="100" velocity="${50*M_PI/180}" lower="${-7.072*M_PI/180}" upper="${37.072*M_PI/180}" />    
    </joint>

    <link name="ring0">
      <inertial>
        <mass value="0.01" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>        
      <visual>
	      <material name="GreyA" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="ring0_geom">
		     <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle1.dae" />
	      </geometry>  
      </visual>      
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="ring0__collision">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle1.stl" />
	      </geometry>
      </collision>
    </link>

    <joint name="right_hand_j7" type="revolute">
      <axis xyz="0 1 0" />
      <limit effort="100" velocity="${1600*M_PI/180}" lower="${-34.99*M_PI/180}" upper="${34.99*M_PI/180}" /> 
      <origin xyz="0 0 -0.044" rpy="0 0 0" />
      <parent link="ring0"/>
      <child link="ring1"/>
    </joint>

    <link name="ring1">    
      <inertial>
        <mass value="0.0429" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>       
      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="ring1_geom">
		      <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle2.dae" />
	      </geometry>  
      </visual>
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="ring1_collision">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle2.stl" />
	      </geometry>
      </collision>
    </link>

    <joint name="right_hand_j8" type="revolute">
      <axis xyz="0 1 0" />
      <limit effort="100" velocity="${1600*M_PI/180}" lower="${-34.21*M_PI/180}" upper="${34.213*M_PI/180}" /> 
      <origin xyz="0 0 -0.032" rpy="0 0 0" />
      <parent link="ring1"/>
      <child link="ring2"/>
    </joint>

    <link name="ring2">
      <inertial>
        <mass value="0.01" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>        
      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="ring2_geom">
		      <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle3.dae" />
	      </geometry>  
      </visual>
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="ring2_collision">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle3.stl" />
	      </geometry>
      </collision>
    </link>

<!-- Start Pinky Finger -->
    <joint name="right_hand_j9" type="revolute">
      <axis xyz="0 1 0" />
      <origin xyz="0 -0.028 -0.101" rpy="0 0 0" />
      <parent link="palm" />
      <child link="pinky0"/>
	<limit effort="100" velocity="${50*M_PI/180}" lower="${-7.072*M_PI/180}" upper="${37.072*M_PI/180}" />    
    </joint>

    <link name="pinky0">
      <inertial>
        <mass value="0.01" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>        
      <visual>
	      <material name="GreyA" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="pinky0_geom">
		     <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle1.dae" />
	      </geometry>  
      </visual>      
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="pinky0__collision">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle1.stl" />
	      </geometry>
      </collision>
    </link>

    <joint name="right_hand_j10" type="revolute">
      <axis xyz="0 1 0" />
      <limit effort="100" velocity="${1600*M_PI/180}" lower="${-34.99*M_PI/180}" upper="${34.99*M_PI/180}" /> 
      <origin xyz="0 0 -0.044" rpy="0 0 0" />
      <parent link="pinky0"/>
      <child link="pinky1"/>
    </joint>

    <link name="pinky1">      
      <inertial>
        <mass value="0.0429" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>     
      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="pinky1_geom">
		      <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle2.dae" />
	      </geometry>  
      </visual>
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="pinky1_collision">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle2.stl" />
	      </geometry>
      </collision>
    </link>

    <joint name="right_hand_j11" type="revolute">
      <axis xyz="0 1 0" />
      <limit effort="100" velocity="${1600*M_PI/180}" lower="${-34.21*M_PI/180}" upper="${34.213*M_PI/180}" /> 
      <origin xyz="0 0 -0.032" rpy="0 0 0" />
      <parent link="pinky1"/>
      <child link="pinky2"/>
    </joint>

    <link name="pinky2">
      <inertial>
        <mass value="0.01" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>        
      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="pinky2_geom">
		      <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle3.dae" />
	      </geometry>  
      </visual>
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="pinky2_collision">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle3.stl" />
	      </geometry>
      </collision>
    </link>
</macro>

  <macro name="m3_r_hand" params="side parent reflect *origin">
    <m3_r_palm side="${side}" reflect="${reflect}" parent="${parent}">
      <insert_block name="origin" />
    </m3_r_palm>
  </macro>


</robot>
